refactor: move index class members from protected to private

jonatack opened the pull request:
Simplify the index base and child classes and improve encapsulation and clarity by making fully overridden virtual member functions `private`. In the child classes, this allows having only public and private members. This change also help contributors add virtual functions into the correct section. Also rename the `GetName()` getters to `GetIndexName()` for clarity and easier grepping/reviewing.

Some context:

A base class virtual function doesn't need to be accessible to be overridden, but it can only be `private` if fully overridden, e.g. no inherited base implementation is used by the child classes.

References:

- https://en.cppreference.com/w/cpp/language/virtual: "Base::vf does not need to be accessible or visible to be overridden. (Base::vf can be declared private, or Base can be inherited using private inheritance.)"

- "An implementation member should be protected if it provides an operation or data that a derived class will need to use in its own implementation. Otherwise, implementation members should be private." - C++ Primer (Lippman, Lajoie, Moo)

- Herb Sutter in http://www.gotw.ca/publications/mill18.htm: "Make virtual functions private, and only if derived classes need to invoke the base implementation of a virtual function, make the virtual function protected. Prefer to make interfaces non-virtual using the Template pattern."

jonatack committed:
scripted-diff: rename GetName() to GetIndexName() in src/index

to improve grepping for a more specific function name.

-BEGIN VERIFY SCRIPT-
sed -i 's/GetName/GetIndexName/g' ./src/index/*
-END VERIFY SCRIPT-

jonatack committed:
index, refactor: move GetIndexName() getters from protected to private

and add doxygen comments

jonatack committed:
index, refactor: move WriteBlock() from protected to private

and add doxygen comments, and touch up doxygen indentation on one line per
review request

jonatack committed:
index, refactor: move GetDB() from protected to private

jonatack committed:
index, refactor: move TxIndex::DB from protected to private

jonatack committed:
index, refactor: move Init(), CommitInternal() and Rewind() to private

PastaPastaPasta reviewed (approved):
utACK 732fafc975aaaabaf9373b8c31cb0952698fee60, it makes sense to me to make stuff private instead of protected where possible

jonatack commented:
Rebased per `git range-diff 7003b6a 732fafc 7785095` and took suggestion in https://github.com/bitcoin/bitcoin/pull/24150#issuecomment-1036954660.

Invalidates previous ACKs by @PastaPastaPasta and @RandyMcMillan; mind re-ACKing?

jonatack commented:
@PastaPastaPasta and @RandyMcMillan, would you mind re-ACKing here?

PastaPastaPasta reviewed (approved):
re-utACK

RandyMcMillan commented:
tACK 77850955b8504638ebcbe8368078c0531ebfc24a


tested on

`Darwin ₿ 19.6.0 Darwin Kernel Version 19.6.0: Sun Nov 14 19:58:51 PST 2021; root:xnu-6153.141.50~1/RELEASE_X86_64 x86_64
`

brunoerg reviewed (approved):
ACK 77850955b8504638ebcbe8368078c0531ebfc24a

Agreed on updating them to `private` and changing from GetName to GetIndexName seems better to understand it.

jonatack commented:
Seems ready for merge.

jonatack committed:
index, refactor: move SetBestBlockIndex() to private

and fix some clang formatting

jonatack commented:
Rebased, and added 4fdd404e652ca6443f666c05341ccb775f461350 left over from a dropped change in #21726. 

@PastaPastaPasta, @RandyMcMillan and @brunoerg, mind re-ACKing? Thanks :heart:

brunoerg reviewed (approved):
re-ACK 4fdd404e652ca6443f666c05341ccb775f461350

laanwj commented on src/index/base.h:
```diff
@@ -78,6 +78,17 @@ class BaseIndex : public CValidationInterface
+    /// Update the internal best block index as well as the prune lock.
+    void SetBestBlockIndex(const CBlockIndex* block);
+
+    /// Get the name of the index for display in logs.
```
I'm kind of confused by this; it's (like `GetDB()`) a method that child classes are supposed to override. What does it mean for it to be private?

jonatack replied:
Thanks for reviewing! It looks to me that a base class virtual method doesn't need to be accessible to be overridden, but it can only be `private` if fully overridden, e.g. no inherited base implementation is used by the child classes.

References:

- https://en.cppreference.com/w/cpp/language/virtual

- "An implementation member should be protected if it provides an operation or data that a derived class will need to use in its own implementation. Otherwise, implementation members should be private." - C++ Primer (Lippman, Lajoie, Moo)

jonatack replied:
Updated the OP with this context.

laanwj replied:
Thanks for looking it up! I have no objection to doing it this way, but it's a dusty corner of C++ to me.

We might want to move the methods that are intended to (or, must) be overridden by child classes together in the class declaration, for documentation purposes.

jonatack replied:
> Thanks for looking it up! I have no objection to doing it this way, but it's a dusty corner of C++ to me.
> 
> We might want to move the methods that are intended to (or, must) be overridden by child classes together in the class declaration, for documentation purposes.

Done!

jonatack committed:
index, refactor: document and group BaseIndex virtual functions

in the class declaration by access and characteristic
for clarity and documentation purposes.

w0xlt reviewed (approved):
crACK https://github.com/bitcoin/bitcoin/pull/24150/commits/f5e985b4b5fcaf6b6341156b00725e8f32373ac4

brunoerg commented:
re-ACK f5e985b4b5fcaf6b6341156b00725e8f32373ac4

jonatack commented:
Updated the motivation in the PR description and simplified the last commit to be move-and-documentation only.

bvbfan commented on src/index/base.h:
```diff
@@ -26,14 +26,37 @@ struct IndexSummary {
 {
+public:
+    /// Destructor interrupts sync thread if running and blocks until it exits.
+    virtual ~BaseIndex();
```
You don't need to mark it virtual, there is a bug and it is destructor of `CValidationInterface` should marked as `virtual`

achow101 commented:
Are you still working on this?

achow101 commented:
Closing this as it has not had any activity in a while. If you are interested in continuing work on this, please leave a comment so that it can be reopened.

jonatack commented:
> Are you still working on this?

Yes. Might be useful to re-open rather than opening a new one, as there is discussion and some feedback to address.

aureleoules commented on src/index/base.h:
```diff
@@ -47,6 +70,14 @@ class BaseIndex : public CValidationInterface
+
+    void ChainStateFlushed(const CBlockLocator& locator) override;
+
+    const CBlockIndex* CurrentIndex() { return m_best_block_index.load(); };
```
```suggestion
    const CBlockIndex* CurrentIndex() const { return m_best_block_index.load(); };
```

aureleoules reviewed (commented):
Concept ACK

fanquake commented:
Moved to draft for now. Was reopened 3 months ago to continue work / follow up on discussion, but hasn't been touched. Has needed rebase for 6 months.
